sentry-core/src/main/java/io/sentry/core/EnvelopeReader.java

```diff
@@ -0,0 +1,211 @@
+    if (header.getEventId() == null || header.getEventId() == SentryId.EMPTY_ID) {
+      throw new IllegalArgumentException("Envelope header is missing required 'event_id'.");
+    }
+
+    int itemHeaderStartOffset = envelopeEndHeaderOffset + 1;
+
+    int payloadEndOffset;
+    List<SentryEnvelopeItem> items = new ArrayList<>();
+    do {
+      int lineBreakIndex = -1;
+      // Look from startHeaderOffset until line break to find next header
+      for (int i = itemHeaderStartOffset; i < envelopeBytes.length; i++) {
+        if (envelopeBytes[i] == '\n') {
+          lineBreakIndex = i;
+          break;
+        }
+      }
+
+      if (lineBreakIndex == -1) {
+        throw new IllegalArgumentException(
+            "Invalid envelope. Item at index '"
+                + items.size()
+                + "'. "
+                + "has no header delimiter.");
+      }
+
+      SentryEnvelopeItemHeader itemHeader =
+          deserializeEnvelopeItemHeader(
+              envelopeBytes, itemHeaderStartOffset, lineBreakIndex - itemHeaderStartOffset);
+
+      if (itemHeader.getLength() <= 0) {
+        throw new IllegalArgumentException(
+            "Item header at index '"
+                + items.size()
+                + "' has an invalid value: '"
+                + itemHeader.getLength()
+                + "'.");
+      }
+
+      payloadEndOffset = lineBreakIndex + itemHeader.getLength() + 1;
+      if (payloadEndOffset > envelopeBytes.length) {
+        throw new IllegalArgumentException(
+            "Invalid length for item at index '"
+                + items.size()
+                + "'. "
+                + "Item is '"
+                + payloadEndOffset
+                + "' bytes. There are '"
+                + envelopeBytes.length
+                + "' in the buffer.");
+      }
+
+      // if 'to' parameter overflows, copyOfRange is happy to pretend nothing happened. Bound need
+      // checking.
+      byte[] envelopeItemBytes =
+          Arrays.copyOfRange(
+              envelopeBytes, lineBreakIndex + 1, payloadEndOffset /* to is exclusive */);
+
+      SentryEnvelopeItem item = new SentryEnvelopeItem(itemHeader, envelopeItemBytes);
+      items.add(item);
+
+      if (payloadEndOffset == envelopeBytes.length) {
+        // End of envelope
+        break;
+      } else if (payloadEndOffset + 1 == envelopeBytes.length) {
+        // Envelope items can be closed with a final line break
+        if (envelopeBytes[payloadEndOffset + 1] == '\n') {
+          break;
+        } else {
+          throw new IllegalArgumentException("Envelope has invalid data following an item.");
+        }
+      }
+
+      itemHeaderStartOffset = payloadEndOffset + 1; // Skip over delimiter
+    } while (true);
+
+    return new SentryEnvelope(header, items);
+  }
+
+  SentryEnvelopeHeader deserializeEnvelopeHeader(byte[] buffer, int offset, int length) {
+    String json = new String(buffer, offset, length, StandardCharsets.UTF_8);
```

Comment:
`StandardCharsets.UTF_8` is API >= 19, did the build work?

do it like our class `HttpTransport.UTF_8`

Reply:
Good catch. Let me check.